text/0002-signature-based-completions.md

```diff
@@ -0,0 +1,206 @@
+
+    git co hello
+
+could be completed to
+
+    git co "hello world"
+
+The span will never include quoting characters and other syntactical elements that aren't considered
+part of the value. Since completers have access to the full line, they can still consider those
+elements, but ideally completers never have to concern themselves with syntax.
+
+# Drawbacks
+
+I see no reason to not do this. Beyond the effort, this approach offers far too many advantages to
+completion than what we currently have, which doesn't feel idiomatic at all since it does not
+tightly integrate with nushell's structured descriptions of commands.
+
+# Rationale and alternatives
+
+The only real alternative would be to continue doing the line-based approach we currently do. This
+has a huge downside of reproducing a lot of the effort that has went into our parser in terms of
+finding out the position of the cursor in terms of flags, positional arguments, etc.
+
+# Prior art
+
+Fish is a shell known for a great completion system. Fish keeps things simple by offering a
+`complete` command that makes it very easy for someone to provide custom completions:
+
+    complete -c myprog -s o -l output -a "yes no"
+
+This provides completions for the command `myprog`. It's pieces are:
+
+  - `-s o`, a short flag named "o";
+  - `-l output`, a long flag named "output"; and
+  - `-a "yes no"`, positional arguments having either a value of "yes" or "no".
+
+This is similar to what nushell offers, minus the fact that nushell expresses these things through
+signatures (note: we have yet to provide an external format for signatures).
+
+Things gets a little more challenging once we get into subcommands:
+
+    complete -c myprog -n "__fish_seen_subcommand_from set-option" -a "(myprog list-options)"
+
+`-n` allows users to configure scripts to determine if the completion should be used. In this case,
+fish has a builtin to determine if a subcommand has been used. In the above example, if the user has
+typed `myprog set-option `, pressing tab will result in completion options being given based on
+whatever `myprog list-options` outputs.
+
+The beauty of a signature based approach in nushell is that subcommands are completely separate from
+their root, so nushell will take care of figuring this out for the user. Each subcommand will have
+its own signature.
+
+# Unresolved questions
+
+None that I can think of.
+
+# Future possibilities
+
+Completions are a huge part of shells, having the potential to provide a ton of productivity to the
+user if sufficiently powerful. The improvements described here will provide a lot of completion
+power to nushell users, but there's likely more room for improvement.
+
+One thing not covered here is how users could specialize a shape. For example, `git checkout <TAB>`
+would make sense to offer various refs - branch names, tags, remote refs, and so on - as completion
```

Comment:
With `git checkout` its even worse, since that can also take files - I guess that's partially why they introduced separate `git switch` and `git restore` commands.

Reply:
Yeah, these completions can get pretty gnarly, that's for sure!

I'm hoping with this work, that'd still be really easy to deal with, because you'd supply two completers for this completion location:

- `git ls-files` 
- `git show-ref`